### ISSUE TITLE
Run Holistic in a Web Worker

### ISSUE THREAD

**Issue by AmitMY** [google/mediapipe#2506]
**System information** (Please provide as much relevant information as possible)
- Have I written custom code (as opposed to using a stock example script provided in Mediapipe): **Yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04, Android 11, iOS 14.4): **Chrome**
- [MediaPipe version](https://github.com/google/mediapipe/releases): **^0.4.1628005088**
- Solution (e.g. FaceMesh, Pose, Holistic): **Holistic**
- Programming Language and version ( e.g. C++, Python, Java): **Javsacript**

**Describe the expected behavior:**
In order to run models on the web without the browser hanging (i.e. without clogging the main thread), it is recommended to use a web worker.

Like other image models, where one can [communicate](https://github.com/sign-language-processing/sign-translate/blob/805139558d533179b3d3b12e3f715b06db1cd0a3/src/app/modules/pix2pix/pix2pix.service.ts#L28) with a [web worker](https://github.com/sign-language-processing/sign-translate/blob/805139558d533179b3d3b12e3f715b06db1cd0a3/src/app/modules/pix2pix/pix2pix.worker.ts), I am trying to get that to work with Holistic.

-----

# Problems:

## 1. Loading path is not respected

Despite setting up the model [this way](https://github.com/sign-language-processing/sign-translate/blob/holistic-worker/src/app/modules/pose/pose.worker.ts#L12):

```ts
model = new Holistic({locateFile: (file) => `/assets/models/holistic/${file}`});
```

While the `js` files are coming from the correct path, the `data`/`wasm` files do not:
![image](https://user-images.githubusercontent.com/5757359/132204407-3c5d22a0-1ff1-441e-aa06-77a54f1c4037.png)

(if performed on the main thread, it performs as expected)

## 2. Passing an image as an OffscreenCanvas fails

It is not possible to pass a `HTMLVideoElement`, or `HTMLCanvasElement` to a web worker, so instead, at first I take the video frame to a canvas and [send its ImageData instead](https://github.com/sign-language-processing/sign-translate/blob/holistic-worker/src/app/modules/pose/pose.service.ts#L36-L43).

Then, in the web worker, I [construct an `OffscreenCanvas`](https://github.com/sign-language-processing/sign-translate/blob/holistic-worker/src/app/modules/pose/pose.worker.ts#L34-L36) and `send` that to the model.

The resulting error is as follows:
![image](https://user-images.githubusercontent.com/5757359/132205118-18650697-1e0b-471f-a3a6-6e76fe58d768.png)

Referring to `v.canvas` in the holistic dist code:
https://unpkg.com/@mediapipe/holistic@0.4.1628005088/holistic.js

**Comment by mhays-google** [comment]
Hey Amit, I don't see anything wrong right off the bat. But the initiator for the requests to the data and wasm files is different than the js files.  `locateFile` actually takes two parameters.  Can you do me a favor and check to see if it's just giving you weird signatures (weirder than what you've posted already, that is)?

```
    this.holisticSolution = new holistic.Holistic({
      locateFile: (path, base) => {
        console.log('path', path, 'base', base);
        return `${base}/${path}`;
      }
    });
```

And good job with the `OffscreenCanvas`! We should be supporting that properly in the near future, since a lot of us want to use worker threads.

**Comment by tyrmullen** [comment]
Just to chime in, reading video on the main thread and sending it over to the worker is definitely the right overall approach for now, although I had a few additional pointers:
- Since you're using Chrome, you can get much better performance by avoiding ImageData, as ImageData forces our data to be transferred on the CPU.  ImageBitmap should be used instead, so all resources stay on the GPU.

- You shouldn't need to pass the image as an OffscreenCanvas-- ImageBitmap should suffice, since it's just being used to create a WebGLTexture anyways-- if you want to use ImageData instead, you can just cast it to another type, like HTMLVideoElement (although you might need to give it 'videoWidth' and 'videoHeight' properties, set equal to 'width' and 'height', respectively).

- However, you're right that OffscreenCanvas is what's missing here-- the JS Solutions framework is currently hard-coded to create an HTMLCanvasElement for internal use, so a patch in MediaPipe code would be necessary to avoid the "canvas undefined" error you're seeing.

**Comment by AmitMY** [comment]
@mhays-google Here is the console output for `path` and `base`:

![image](https://user-images.githubusercontent.com/5757359/132715850-cf6cde6d-0d08-4275-9869-fffee1a0cc76.png)


@tyrmullen Thanks, I am now doing `const bitmap = await createImageBitmap(video);` and passing that to the worker.
I could either draw it on an `OffscreenCanvas` or just pass it directly to Holistic, however, both are not currently supported because, like you said, it creates a `canvas` which is not available in the worker context.

**Comment by tvanier** [comment]
@AmitMY for the loading path, maybe try the `URL` constructor with `origin` as second argument?
```js
model = new Holistic({
  locateFile: (file) =>
    new URL(`/assets/models/holistic/${file}`, globalThis.location.origin).toString()
});
```

**Comment by tvanier** [comment]
and +1 for a lower-level API (ImageBitmap) with support for web workers (I use selfie-segmentation)
where should I upvote? ;-)

**Comment by AlexShafir** [comment]
I found a temporary fix for this problem: 
1. Beatify holistic.js
2. Add this code after holistic.js line 993: `d.C = d.h.GL.currentContext.GLctx`
3. Put your worker file in the same folder with holistic.js
4. For every message you post to worker, do not send post new messages before you receive result.

**Comment by AmitMY** [comment]
Thanks for the workaround, @AlexShafir 
However, @sgowroji, I still argue for a native solution in the package, and support for `OffscreenCanvas` or better yet, an `ImageBitmap`

**Comment by AlexShafir** [comment]
@AmitMY  ImageBitmap is supported by holistic solution, you can see it in my demo.
index.js:
```js
createImageBitmap(videoElement).then((bitmap) => {
        holisticWorker.postMessage(bitmap, [bitmap]) // transferable
    })
```

holistic_worker.js:
```js
onmessage = (event) => {
    holistic.send({image: event.data})
}
```

**Comment by AlexShafir** [comment]
For me bummer is that after workaround holistic.js is ultra slow even on high-end device, so I switched to face_mesh.js which shows a lot better performance.

**Comment by AmitMY** [comment]
> ```js
> new URL(`/assets/models/holistic/${file}`, globalThis.location.origin).toString()
> ```

Thanks, @tvanier , this does not change the result, unfortunately. The path is still not correctly resolved

**Comment by mhays-google** [comment]
Workers are on our roadmap, but I unfortunately don't have a window to
provide this.  ImageBitmap is already an output format where supported
(i.e., Chrome), but if I hear you right, you would like an input format of
ImageBitmap?

**Comment by tyrmullen** [comment]
ImageBitmap is already an input format as well, as demonstrated by @AlexShafir .

**Comment by AmitMY** [comment]
@AlexShafir Your changes seem to be minimal. 
If it's not too much trouble, could you please contribute a fix to this repository?

**Comment by AlexShafir** [comment]
@AmitMY Source code for Js Solutions is not released, see #1408 (basically wasm does not fit nicely into Bazel). So one cannot make PR.

My solution is just a workaround. To make it complete, one needs to add proper path handling - at least.

**Comment by AmitMY** [comment]
Thanks @tyrmullen 
I'll just note that the largest, and most apparent, and annoying main thread block is *loading* the holistic model, not inference, in my case.
When it is loading, nothing else in my app works, and that sucks because it happens at a time when people do want to interact with the app. That's my main use case here, and secondary is inference block.

**Comment by tyrmullen** [comment]
No problem!  And yes, that makes sense, and workers could probably help for that-- I'd be curious to see just how helpful they can be on this front (they should allow us to make those "wait for GPU" ops be non-blocking, but would still impact the GPU queue).

One thing to note is that loading the model data is likely taking very little time itself (or it'd show as CPU processing time), but rather the initialization of the ML inference engine is quite expensive.  I suspect the concentration of loading time you're seeing will be spread between two spikes: (a) one occurring during the MediaPipe graph initialization and (b) one occurring during the first frame processed (the first 'send' call), probably right afterwards.  That's because shader program instantiation is likely the main cost here, and WebGL lazily enforces shader initialization, so while a lot of work is done up front, the process doesn't complete until the first frame is rendered.  If your performance profiling matches this pattern, then this is almost certainly the culprit.

**Comment by AmitMY** [comment]
Can confirm this is still an issue in version `0.5.1635989137`

**Comment by fgsreally** [comment]
My personal feeling is, pose Initialize () does not complete the real initialization. One trick is to send a non empty video element in advance before playing the video. However, a moment of Caton cannot be avoided in any way (including worker).

I hope there is a way to reduce the impact of real initialization on the window. I mean the blocking of interface rendering rather than time-consuming. Even if it takes more time, it is worth it (because there are ways to cover up the disadvantage of time-consuming, but there is no way to solve the blocking of rendering

**Comment by fgsreally** [comment]
My personal feeling is, pose Initialize () does not complete the real initialization. One trick is to send a non empty video element in advance before playing the video. However, a moment of Caton cannot be avoided in any way (including worker).

I hope there is a way to reduce the impact of real initialization on the window. I mean the blocking of interface rendering rather than time-consuming. Even if it takes more time, it is worth it (because there are ways to cover up the disadvantage of time-consuming, but there is no way to solve the blocking of rendering

Web worker is not the key

**Comment by tvanier** [comment]
I second @AmitMY , it would be great if the mediapipe JS could run in a worker, and `send` resolving with the actual results.
Would fit nicely with the [insertable streams API](https://web.dev/mediastreamtrack-insertable-media-processing/)!

```js
// worker.js
const holistic = new Holistic(/* ... */)
const canvas = new OffscreenCanvas(/* ... */)

const transformer = new TransformStream({
    async transform(videoFrame, controller) {
        const results = await holistic.send({ image: videoFrame })
        // draw results onto canvas and
        const newFrame = new VideoFrame(canvas)
        controller.enqueue(newFrame)
    }
})

self.onmessage = (message) => {
    // main thread transfers a MediaStreamTrackProcessor.readable and MediaStreamTrackGenerator.writable
    const { readable, writable } = message.data
    readable.pipeThrough(transformer).pipeTo(writable)
}
```

**Comment by w-okada** [comment]
I created the webworker demo using the mediapipe models (face, hand, pose).
https://d3iwgbxa9wipu8.cloudfront.net/P01_wokers/t18_mediapipe-mix/index.html

(Note: some part of process is guess-work.)
The repo is linked at the upper right icon.

**Comment by AmitMY** [comment]
@w-okada Seems like you are using `@tensorflow-models` and not `mediapipe` directly. Please correct me if I am mistaken.
While that can work, this issue is specifically about running `mediapipe` models.

**Comment by w-okada** [comment]
@AmitMY 
I use the model listed at the page below.
https://google.github.io/mediapipe/solutions/models.html

**Comment by ButzYung** [comment]
@tyrmullen  Running mediapipe on web worker may not do much on high-end devices, but it does make a difference on lower-end devices. Say if your device can only run mediapipe at below 20fps, if you run everything on the main thread, your graphics rendering will also be stuck at the same frame rate (the cause can be the CPU overhead used by the mediapipe models, or it's the way the GPU resources are shared in the main thread). However, if you put mediapipe on a web worker, you can keep your main thread rendering at 30fps or higher with very little impact on the inference speed of the mediapipe models.

I have a web app that loads mediapipe on web workers. I put a stress test on a medicore smart phone, and although the inference speed is just around 5fps, the 3D graphics on the main thread is still at 30fps.

https://twitter.com/butz_yung/status/1519365509128417280/

**Comment by tyrmullen** [comment]
@ButzYung That's what I meant by saying workers wouldn't help pure performance of the ML inference-- that the 5fps framerate would not increase (and might potentially decrease, depending).  Workers can definitely help unblock the main thread, for fast and slow devices alike, but when it's GPU it's usually less about overhead and more about synchronization and waiting; Holistic has to wait on CPU for GPU to finish for every frame, while most pure rendering doesn't (and shouldn't).  So that's why I said running on workers could definitely help overall app behavior, but not the inference framerate (or initialization/load time) [all this is assuming inference is running on GPU-- on CPU you get actual gains, often 2-3x factor, since the threads are relatively independent].

However, all that being said, I'm a bit surprised/impressed that you see so little impact to your main rendering thread given that I'd imagine the GPU should be quite fully utilized by the ML inference (if that part can only run at 5fps-- which is also quite low for GPU ML, even on moderate devices); that makes me suspect the inference is running on CPU and not GPU.  Were you testing on an iPhone by any chance? (we force CPU inference there).  I guess one way to check too would be to look at the Chrome Developer console performance traces and see if the worker looks like it's using a lot of WebGL calls at the end of the call stacks?

**Comment by KTRosenberg** [comment]
> @AmitMY I took a similar approach as @AlexShafir did before by editing the beautified version of holistic.js (BTW I tried his method on `0.5.1635989137` some hours ago but it didn't seem to work anymore). Our edits are basically on the same location.
> 
> ```
>             case 6:
>                 e.h = p.h;
> // hack for worker
> /*
>                 a.l = new OffscreenCanvas(1,1);
>                 a.h.canvas = a.l;
>                 g = a.h.GL.createContext(a.l, {
>                     antialias: !1,
>                     alpha: !1,
>                     aa: "undefined" !== typeof WebGL2RenderingContext ? 2 : 1
>                 });
>                 a.h.GL.makeContextCurrent(g);
>                 p.g = 4;
>                 break;
> */
>             case 7:
> // hack for worker
>                 a.l = ((typeof Document !== 'undefined') && document.createElement("canvas")) || new OffscreenCanvas(1,1);
> ```
> 
> My approach is to basically comment out the whole case 6, and do everything worker and non-worker on case 7. And on the first line, I edit it to use `OffscreenCanvas` whenever web worker is used.
> 
> As for the path issue, put the worker file in the same folder with holistic.js, just like AlexShafir's version did.

So you need the offscreen canvas? It would be nice to have a complete running example available.

**Comment by ButzYung** [comment]
@tyrmullen I tested it on an android phone with Snapdragon 845. I suppose it is running on GPU (I doubt it will be anything more than 1fps if it runs on CPU), but unfortunately I can't check the web console with my current setup. In fact, I tested it with non-worker case as well, and the inference speed is almost the same (5-7fps) but the 3D rendering speed drops significantly to around the same fps as the inference.

As to why 3D rendering speed can be maintained when ML model is running on web worker, we make the wrong assumption that ML inference can fully utilize the GPU, but in fact it cannot. I tested my app on a medicore PC (GTX1650) with the inference speed being around 25fps average, but the ML inference and 3D rendering combined is using less than 1/3 of the the total GPU speed. There should be enough GPU resource to improve the inference speed, but for some reasons the browser just leaves it idle. That's why the 3D rendering speed in the main thread can be maintained and they won't affect each other (as long as neither side is using too much GPU resource).

PS: To anyone curious, you can test my web app here.

worker enabled:
https://sao.animetheme.com/XR_Animator.html

worker disabled:
https://sao.animetheme.com/XR_Animator.html?ML_worker_disabled=1

**Comment by tyrmullen** [comment]
It's been a while since I've tried anything on Android Chrome (I'm assuming this is Chrome), so that's very good to know!  I wonder if that's because the repeated CPU/GPU sync-ing gives it some breathing room, or if the browser is playing a more active role in the throttling.  Perhaps one way to test the first would be to try with Segmentation instead of Holistic and see if that changes things, since Segmentation doesn't have any CPU/GPU syncs (Holistic has one in order to bring back the GPU ML to CPU so the pose landmarks can be handled CPU-side)?

With the PC experiments you'll have to be a little bit more careful, since if you're using webcam, the browser definitely caps the framerate to 25-30fps, so it sounds like it could possibly be that you're just hitting that framerate limit, and won't be able to really test GPU utilization fully.  Also, your PC's overall GPU situation should definitely always be underutilized, since there should be a single shared-across-all-tabs Chrome GPU process, so if you look at more fine-grained device metrics, I'd expect it to look like one GPU thread is being utilized a ton (for all the 3D rendering and ML inference), while the other threads may likely be sitting idle or not doing much.  But I haven't checked out how utilized that one thread is (and haven't really been running on Android lately), and at least on Android with Snapdragon 845, it sounds like perhaps surprisingly the answer is "not fully utilized at all", which runs counter to a lot of our earlier PC experiments.

In any case, definitely something interesting to look into and keep in mind.

**Comment by Justinyu1618** [comment]
@ButzYung Did you have to edit any other part of the code to get this to work? When I try to initialize the model in a web worker, it tries to run `importScripts` on the tflite file, which errors out, in addition to a bunch of other issues because there's no `document` or `window`.

**Comment by ButzYung** [comment]
@Justinyu1618 Yeah you have to edit `holistic.js`, as shown in my previous comment. Errors should be gone after that.

**Comment by Wei-1** [comment]
@ButzYung 's editing actually work, here is a simpler working example:
https://github.com/Wei-1/holistic-website-test

**Comment by davideloba** [comment]
@Wei-1 thank you for your working example! I'm trying to add the same fix to the facemesh.js but I found another problem there. The switch doesn't reach the case 6 but it stucks into the case 2, importScripts works but it's called over and over again. This problem isn't present in your holistic.js..

```
                    case 2:
                        .........
                        importScripts.apply(null, L(h));
                        return N(m, createMediapipeSolutionsWasm(Module), 6);
                    case 6: <---- never reached
                        d.h = m.h;
                    // d.l = new OffscreenCanvas(1, 1);
```

**Comment by AmitMY** [comment]
@mhays-google as you may see there is a solution - 
could you please implement this in the public npm release instead of people needing to edit themselves?

having it running in a worker thread is essential for performant and fluent user experience

**Comment by davideloba** [comment]
thanks to @Wei-1 and @ButzYung 
[this](https://github.com/davideloba/facemesh) is my test with patched facemesh

**Comment by ayoub-epsano** [comment]
any solution for this issue ?

**Comment by x6ud** [comment]
I found that pose's web worker support implement is incomplete. Here is my solution without modifying the origin file:

1. Copy `/node_modules/@mediapipe/pose` to a static path, in my case is `/assets`

2. Create a classic web worker (rather than module type, otherwise you can't use importScripts)
`
const worker = new Worker('/assets/detect-pose.worker.js', {type: 'classic'});
`

3. In `detect-pose.worker.js`

```javascript
const POSE_CONFIG = {
    locateFile(path, prefix) {
        return '/assets/@mediapipe/pose/' + path;
    }
};
self.createMediapipeSolutionsWasm = POSE_CONFIG;
self.createMediapipeSolutionsPackedAssets = POSE_CONFIG;
importScripts(
    '/assets/@mediapipe/pose/pose.js',
    '/assets/@mediapipe/pose/pose_solution_packed_assets_loader.js',
    '/assets/@mediapipe/pose/pose_solution_simd_wasm_bin.js',
);

(function () {
    let pose;
    let detectPoseResults;
    self.onmessage = async function (e) {
        if (!pose) {
            pose = new Pose(POSE_CONFIG);
            pose.setOptions({
                selfieMode: false,
                modelComplexity: 2,
                smoothLandmarks: false
            });

            const solution = pose.g;
            const solutionConfig = solution.g;
            solutionConfig.files = () => []; // disable default import files behavior
            await pose.initialize();
            solution.D = solution.h.GL.currentContext.GLctx; // set gl ctx

            // load data files
            const files = solution.F;
            files['pose_landmark_heavy.tflite'] = (await fetch('/assets/@mediapipe/pose/pose_landmark_heavy.tflite')).arrayBuffer();
            files['pose_web.binarypb'] = (await fetch('/assets/@mediapipe/pose/pose_web.binarypb')).arrayBuffer();

            // set callback
            pose.onResults(function onResults(results) {
                detectPoseResults = results;
            });
        }
        pose.reset();
        const bitmap = e.data;
        await pose.send({image: bitmap});

        postMessage(detectPoseResults);
    };
})();
```

4. To use
```typescript
function detectPose(image: HTMLImageElement) {
    return new Promise<PoseResults>(async function (resolve) {
        worker.onmessage = function (e) {
            resolve(e.data);
        };
        worker.postMessage(await createImageBitmap(image));
    });
}
```

The problem is that `initialize()` is trying to load the data files with `importScripts` in web worker scope, and forgot to set gl to it's context. So just disable the original file importing and load the required files yourself.

**Comment by ayoub-root** [comment]
> 

thanks for your efforts, do you have a working example.?

**Comment by ayoub-root** [comment]
thanks to @x6ud 

the code is not working in react , i made some changes

# create the worker file   in assets  folder
/* eslint-disable no-undef */
/* eslint-disable no-restricted-globals */
const POSE_CONFIG = {
  locateFile(path, prefix) {
    return "/pose/" + path;
  },
};
self.createMediapipeSolutionsWasm = POSE_CONFIG;
self.createMediapipeSolutionsPackedAssets = POSE_CONFIG;
importScripts(
  "/pose/pose.js",
  "/pose/pose_solution_packed_assets_loader.js",
  "/pose/pose_solution_simd_wasm_bin.js"
);
let pose;

(async function () {
  if (!pose) {
    console.log("start pose");
    pose = new Pose(POSE_CONFIG);
    pose.setOptions({
      selfieMode: false,
      modelComplexity: 0,
      smoothLandmarks: false,
    });
    const solution = pose.g;
    const solutionConfig = solution.g;
    solutionConfig.files = () => []; // disable default import files behavior
    await pose.initialize();
    solution.D = solution.h.GL.currentContext.GLctx; // set gl ctx

    // load data files
    const files = solution.F;
    files["pose_landmark_lite.tflite"] = (
      await fetch(" /pose/pose_landmark_lite.tflite")
    ).arrayBuffer();
    files["pose_web.binarypb"] = (
      await fetch(" /pose/pose_web.binarypb")
    ).arrayBuffer();

    // set callback
    pose.onResults(function onResults(results) {
      postMessage(results);
    });
  }
  self.onmessage = async function (e) {
    // const bitmap = e.data;
    await pose.send({ image: e.data });
  };
  pose.reset();
})();

# my pose folder was in /public (i'm working with react) 
#  i updated  this instruction in pose.js (worker doesn't support "document")
 (navigator.userAgent.includes("Mac") && "ontouchend" in document), =>  (navigator.userAgent.includes("Mac")

, i noticed that worker is very slow , (5s and more of latency between the original video"camera'  and the generated one )

**Comment by AmitMY** [comment]
I have tried @x6ud's solution with `@mediapipe/holistic`

**It is very promising, but not yet fully working.**
It seems to be running the pose estimation, but only returns `{image: ImageBitmap, multiFaceGeometry: Array(0)}` instead of the full object with all the poses.

**Code:** https://github.com/sign/translate/pull/28
**Demo:** https://translate-sign-mt--pr28-holistic-worker-nzqu5dt3.web.app/playground

Open console: it will print the `results` for every frame.
![image](https://user-images.githubusercontent.com/5757359/213201968-48a6101a-7396-4458-a7dc-b177fb5c39c7.png)

**Comment by x6ud** [comment]
For more details, the pose.js without minification should be like:
```javascript
async function fetchFile(solution, url) {
    const files = solution.F;
    if (url in files) {
        return files[url];
    }
    url = solution.locateFile(url, '');
    const data = (await fetch(url)).arrayBuffer();
    return files[url] = data;
}

class Pose {
    async initialize() {
        const solution = this.g;
        const config = solution.g;
        let files;
        if (config.files) {
            if (typeof config.files === 'function') {
                let options = solution.j;
                files = config.files(options);
            } else {
                files = config.files;
            }
        } else {
            files = [];
        }
        if (typeof window === 'object') {
            window.createMediapipeSolutionsWasm = {locateFile: solution.locateFile};
            window.createMediapipeSolutionsPackedAssets = {locateFile: solution.locateFile};
            const dataFiles = files.filter(file => file.data);
            const scriptFiles = files.filter(file => !file.data);
            const fetchDataFilesPromise = Promise.all(dataFiles.map(async file => {
                const dataPromise = await fetchFile(solution, file.url);
                if (file.path) {
                    solution.overrideFile(file.path, await dataPromise);
                }
                return dataPromise;
            }));
            const importScriptsPromise = Promise.all(
                scriptFiles.map(async file => {
                    if (file.simd === useSimd) {
                        const basePath = solution.S;
                        await insertScriptTagToDocument(solution.locateFile(file.url, basePath));
                    }
                }))
                .then(async () => {
                    // pose_solution_simd_wasm_bin.js and pose_solution_packed_assets_loader.js will set those funtions to window
                    const wasmCtx = await window.createMediapipeSolutionsWasm(window.createMediapipeSolutionsPackedAssets());
                    solution.h = wasmCtx;
                });
            const fetchGraphPromise = (async () => {
                if (config.graph?.url) {
                    await fetchFile(solution, config.graph.url);
                }
            })();
            await Promise.all([importScriptsPromise, fetchDataFilesPromise, fetchGraphPromise]);
            const canvas = solution.l = document.createElement('canvas');
            let gl = canvas.getContext('webgl2', {});
            if (!gl) {
                gl = canvas.getContext('webgl', {});
            }
            if (!gl) {
                alert('error');
                return;
            }
            solution.D = gl;
            const wasmCtx = solution.h;
            wasmCtx.canvas = canvas;
            wasmCtx.createContext(canvas, true, true, {});
        } else {
            // run in web worker
            if (typeof importScripts !== 'function') {
                throw Error('...');
            }
            const fileUrls = files.filter(file => file.simd === useSimd).map(file => {
                const basePath = solution.S;
                return solution.locateFile(file.url, basePath);
            });
            // this will cause an error when importing .tflite file
            importScripts.apply(null, fileUrls);
            await createMediapipeSolutionsWasm(Module);
            const canvas = solution.l = new OffscreenCanvas(1, 1);
            const wasmCtx = solution.h;
            wasmCtx.canvas = canvas;
            const gl = wasmCtx.createContext(canvas, {antialias: false, alpha: false});
            wasmCtx.GL.makeContextCurrent(gl);
        }
        // ...
    }

    async send(inputs, at) {
        const solution = this.g;
        if (!solution.inputs) {
            return;
        }
        const timestamp = 1e3 * (at == null ? performance.now() : at);
        await this.initialize();
        const packetDataList = new PacketDataList();
        for (let key of Object.keys(inputs)) {
            const input = inputs[key];
            const info = solution.inputs[key];
            switch (info.type) {
                case 'video': {
                    const cachedTexture = this.m;
                    let texture = cachedTexture[info.stream];
                    if (!texture) {
                        const wasmCtx = solution.h;
                        const gl = solution.D;
                        texture = cachedTexture[info.stream] = new Texture(wasmCtx, gl);
                        const glTexture = texture.l;
                        if (glTexture === 0) {
                            texture.l = wasmCtx.createTexture();
                        }
                    }
                    let width, height;
                    if (typeof HTMLVideoElement !== 'undefined' && input instanceof HTMLVideoElement) {
                        width = input.videoWidth;
                        height = input.videoHeight;
                    } else if (typeof HTMLImageElement !== 'undefined' && input instanceof HTMLImageElement) {
                        width = input.naturalWidth;
                        height = input.naturalHeight;
                    } else {
                        // you might want to modify this for different transferable objects
                        width = input.width;
                        height = input.height;
                    }
                    const tex = {glName: texture.l, width, height};
                    const gl = texture.g;
                    gl.canvas.width = tex.width;
                    gl.canvas.height = tex.height;
                    gl.activeTexture(gl.TEXTURE0);
                    const wasmCtx = texture.h;
                    wasmCtx.bindTexture2d(texture.l);
                    gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, gl.RGBA, gl.UNSIGNED_BYTE, input);
                    gl.bindTexture2d(0);
                }
                break;
                case 'detections': {
                    // ...
                }
                break;
            }
            // ...
        }
        // ...
    }
}
```

Other solutions may be different.

In addition, you may need to use other [transferable object](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Transferable_objects) to avoid unnecessary data type conversion

**Comment by ButzYung** [comment]
@Saintpreston I haven't tried the new tasks version of the pose landmarker since it is still WIP, but judging from the error message, a wild guess is that you need to load the scripts in the non-module mode of web worker I suppose?

**Comment by ButzYung** [comment]
> Just btw, I am only able to get the trick that @ButzYung shared work up to version: [0.5.1635989137](https://www.npmjs.com/package/@mediapipe/holistic/v/0.5.1635989137) For version: [0.5.1675471629](https://www.npmjs.com/package/@mediapipe/holistic/v/0.5.1675471629) that was updated 3 months ago, I am not able to make it run in the worker. If @Saintpreston you are trying now, maybe use the version [0.5.1635989137](https://www.npmjs.com/package/@mediapipe/holistic/v/0.5.1635989137), which should be easier to start.

`0.5.1675471629` works for me on worker with the same trick. The issue is that the output result of `world_landmarks` has changed from `.ea` to `.za`

**Comment by kuaashish** [comment]
@AmitMY,

MP Tasks for Web support Web Workers. However, We are working towards publish the Holistic solutions Asap from our end And let you know once becomes available.

**Comment by rummik** [comment]
@kuaashish Has this been fixed?  It looks like github-actions just mistakenly closed it and it wasn't reopened afterwards, which might make it a bit difficult to be notified when the issue is actually resolved 😅

**Comment by Wei-1** [comment]
This is no resolved. But if the GitHub action will close any issue after 7 days without action, it will be closed after 7 days if we reopen it anyway?

**Comment by AmitMY** [comment]
Release v0.10.5 *claims*: 
> Don't access "document" in WebWorker

Worth checking now.

**Comment by lc-spxl** [comment]
Holistic is not part of tasks-vision yet. Any  plan for release ? Thank you

**Comment by kyasuda2003** [comment]
> 
> MP Tasks for Web support Web Workers. However, We are working towards publish the Holistic solutions Asap from our end And let you know once becomes available.

Maybe I am bit outdated..

The usage from your [latest doc](https://developers.google.com/mediapipe/solutions/vision/object_detector/web_js)
```javascript
const vision = await FilesetResolver.forVisionTasks(
  // path/to/wasm/root
  "https://cdn.jsdelivr.net/npm/@mediapipe/tasks-vision@latest/wasm"
);
objectDetector = await ObjectDetector.createFromOptions(vision, {
  baseOptions: {
    modelAssetPath: `https://storage.googleapis.com/mediapipe-tasks/object_detector/efficientdet_lite0_uint8.tflite`
  },
  scoreThreshold: 0.5,
  runningMode: runningMode
});
```
if as you suggested the latest MP can be run in a worker, then I assume it can only be run with ```{type: `module`}```

but this signature also suggest ```importScripts``` must be in place for loading the wasm which is ambiguous to yours.
https://github.com/google/mediapipe/blob/master/mediapipe/web/graph_runner/graph_runner.ts#L53

I overloaded the signature to run that in a worker but the performance is less than expected. Is there something you can advise otherwise? [kuaashish](https://github.com/kuaashish)

**Comment by tyrmullen** [comment]
@kyasuda2003 The model you're using there has a `_uint8` suffix, so it's quantized.  This will help it run more efficiently on CPU, but means that it cannot run on GPU, so in isolation it is expected to run much slower than if you were able to use `delegate: GPU`.  I'm not sure why the function `importScripts` is declared there (perhaps just to get rid of some TS compiler warning/error depending on settings?), but in the call where it's actually used (which looks like it's now in `run_script_helper.ts.template`, it is supposed to be used only where it actually exists (i.e. from a worker context), so new MP Solutions in general are expected to run from a worker (on non-iOS platforms) as well as on the main/UI thread (on all supported platforms).

I can also confirm that Holistic is not yet an offering for the new MP Solutions, but cannot comment as to release plans for it.

**Comment by epleaner** [comment]
Bumping this. Has anyone successfully run Holistic in a Web Worker? It's unclear from the earlier comments.

**Comment by epleaner** [comment]
> @epleaner yes with the newest version of Mediapipe, but be aware of the memory. I had some cases of out of the memory in different browser versions. But I think could be my code not cleaning correctly the Offscreencanvas.
> 
> Also, check the module bundler you will use considering not all handle workers correctly. Or just `importScripts` from a CDN.
> 
> Edit: the tflite, js and wasm latest versions are not compatible with previous versions

Do you have any code snippets or a repository that has the newest version working in a Web Worker?

**Comment by Wei-1** [comment]
> > @epleaner yes with the newest version of Mediapipe, but be aware of the memory. I had some cases of out of the memory in different browser versions. But I think could be my code not cleaning correctly the Offscreencanvas.
> > Also, check the module bundler you will use considering not all handle workers correctly. Or just `importScripts` from a CDN.
> > Edit: the tflite, js and wasm latest versions are not compatible with previous versions
> 
> Do you have any code snippets or a repository that has the newest version working in a Web Worker?

Maybe you can try this?
https://github.com/Wei-1/holistic-website-test

**Comment by epleaner** [comment]
Thanks, that does work locally. I'm trying to port this over to NextJS, and oddly I'm getting a whole bunch of window/document is not defined errors using the same holistic.js / Web Worker file in this repository

**Comment by kuaashish** [comment]
Hi @AmitMY,

We have upgraded our MediaPipe Legacy Solutions to new MediaPipe solutions, which now include the new Holistic Task API and supports the Web Worker.

Could you please try using the new [HolisticLandmarker task APIs](https://github.com/google/mediapipe/blob/master/mediapipe/tasks/web/vision/holistic_landmarker/holistic_landmarker.ts), and let us know if the problem persists?

Thank you!!

**Comment by AmitMY** [comment]
Thanks @kuaashish 
The guide says "Coming soon"
https://developers.google.com/mediapipe/solutions/vision/holistic_landmarker

There's also no information in https://www.npmjs.com/package/@mediapipe/tasks-vision

**Comment by kuaashish** [comment]
Hi @AmitMY,

We expect to release samples soon, starting with Android, followed by others. Documentation will be available afterwards. Though we can not give an exact date, it is in progress and will be out soon.

Thank you!!

**Comment by AmitMY** [comment]
(commenting to un-stale the issue, waiting for the solution)

**Comment by stormsc1** [comment]
I am running into the same issues. Do @kuaashish/@schmidt-sebastian know if there is any working examples of the "new" HolisticLandmarker? I can only find CodePens using the old one.

**Comment by Wei-1** [comment]
After all these years, I am still waiting for this.

**Comment by schmidt-sebastian** [comment]
This is now supported. https://github.com/google-ai-edge/mediapipe-samples-web/blob/main/src/tasks/holistic-landmarker.ts and https://github.com/google-ai-edge/mediapipe-samples-web/blob/main/src/workers/holistic-landmarker.worker.ts show an example.
